### PR TITLE
libssh: add recipe

### PR DESCRIPTION
**Pull request by obnyis**
Related: #13449

Related: #16899

Fixes: #16747

Specify library name and version:  **libssh/0.10.6**

Changed replace_in_file from previous PR with apply_conandata_patches

Added `tc.cache_variables["CMAKE_TRY_COMPILE_CONFIGURATION"]` to support Visual Studio / Ninja Multi-Config generators

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/libssh/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        tc.variables["WITH_GCRYPT"] = self.options.crypto_backend == "gcrypt"
+        tc.variables["WITH_GSSAPI"] = False
+        tc.variables["WITH_MBEDTLS"] = False
+        tc.variables["WITH_NACL"] = False
+        tc.variables["WITH_SYMBOL_VERSIONING"] = False
+        tc.variables["WITH_ZLIB"] = self.options.with_zlib
+        if is_msvc(self):
+            tc.variables["USE_MSVC_RUNTIME_LIBRARY_DLL"] = not is_msvc_static_runtime(self)
+        tc.cache_variables["CMAKE_POLICY_DEFAULT_CMP0077"] = "NEW"
+        tc.cache_variables["CMAKE_TRY_COMPILE_CONFIGURATION"] = str(self.settings.build_type)
```
Comment (valgur) [inline recipes/libssh/all/conanfile.py:74]: Nice! This indeed seems to be the correct workaround for this CMake/Conan issue:  https://github.com/conan-io/conan/issues/12180

File: recipes/libssh/all/test_v1_package/CMakeLists.txt
Comment (valgur) [inline recipes/libssh/all/test_v1_package/CMakeLists.txt:1]: New recipes should avoid including `test_v1_package`. You can drop it entirely.

File: recipes/libssh/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    def package_info(self):
+        self.cpp_info.libs = ["ssh"]
+        if not self.options.shared:
+            self.cpp_info.defines.append("LIBSSH_STATIC=ON")
+        if self.settings.os == "Windows":
+            self.cpp_info.system_libs.extend(["ws2_32", "iphlpapi"])
+
+        self.cpp_info.set_property("cmake_file_name", "libssh")
+        self.cpp_info.set_property("cmake_target_name", "ssh")
+        self.cpp_info.set_property("pkg_config_name", "libssh")
```
Comment (valgur) [inline recipes/libssh/all/conanfile.py:104]:
```suggestion
        self.cpp_info.set_property("cmake_file_name", "libssh")
        self.cpp_info.set_property("cmake_target_name", "ssh::ssh")
        self.cpp_info.set_property("cmake_target_aliases", ["ssh"])
        self.cpp_info.set_property("pkg_config_name", "libssh")
```
The library exports both `ssh` and `ssh::ssh` as targets: https://git.libssh.org/projects/libssh.git/tree/src/CMakeLists.txt#n345

The rest are correct:
cmake config file: https://git.libssh.org/projects/libssh.git/tree/src/CMakeLists.txt#n387
.pc file: https://git.libssh.org/projects/libssh.git/tree/CMakeLists.txt#n102

You could also add `ssh-static` and `ssh::static` aliases if not self.options.shared: https://git.libssh.org/projects/libssh.git/tree/src/CMakeLists.txt#n407

**Review by valgur — CHANGES_REQUESTED** [review]
LGTM, should only add a missing exported target alias. Thanks!

File: recipes/libssh/all/conanfile.py (hunk 2 above)
Comment (valgur) [reply recipes/libssh/all/conanfile.py:104]: Including one or more of the above links as a reference in a comment would not hurt either.

File: recipes/libssh/all/test_v1_package/CMakeLists.txt
Comment (obnyis) [reply recipes/libssh/all/test_v1_package/CMakeLists.txt:1]: Neat, I've removed that now

**Review by valgur — APPROVED** [review]
Thanks!

File: recipes/libssh/all/conanfile.py (hunk 2 above)
Comment (obnyis) [reply recipes/libssh/all/conanfile.py]: I added the missing targets, but I'm not sure which lines in particular you want comments on. Could you make a suggested change with what you are after and I'll add it in?

Comment (obnyis) [reply recipes/libssh/all/conanfile.py]: Hi @valgur, I added some comments with links to the lines these definitions are in the libssh-0.10.6 tag as comments. Can you review this and let me know if it's what you had in mind?

**Review by uilianries — COMMENTED** [review]
Looking good, only few details

File: recipes/libssh/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+            ["ssh"] if self.options.shared else ["ssh", "ssh-static", "ssh::static"],
+        )
+        # pkg-config defined at https://git.libssh.org/projects/libssh.git/tree/CMakeLists.txt?h=libssh-0.10.6#n124
+        self.cpp_info.set_property("pkg_config_name", "libssh")
+
+        # TODO: to remove in conan v2 once cmake_find_package_* generators removed
+        self.cpp_info.filenames["cmake_find_package"] = "libssh"
+        self.cpp_info.filenames["cmake_find_package_multi"] = "libssh"
+        self.cpp_info.names["cmake_find_package"] = "libssh"
+        self.cpp_info.names["cmake_find_package_multi"] = "libssh"
```
[the remark this answers is not in the document]
Comment (obnyis) [reply recipes/libssh/all/conanfile.py]: Done. Thanks for the review :)

File: recipes/libssh/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        get(self, **self.conan_data["sources"][self.version], strip_root=True)
+
+    def generate(self):
+        tc = CMakeToolchain(self)
+        tc.variables["CLIENT_TESTING"] = False
+        tc.variables["SERVER_TESTING"] = False
+        tc.variables["WITH_EXAMPLES"] = False
+        tc.variables["WITH_GCRYPT"] = self.options.crypto_backend == "gcrypt"
+        tc.variables["WITH_GSSAPI"] = False
+        tc.variables["WITH_MBEDTLS"] = False
```
Comment (uilianries) [inline recipes/libssh/all/conanfile.py:70]: quick question here, did you try to use the mbedtls package here? https://conan.io/center/recipes/mbedtls?version=3.6.0

It's not a blocker, but I would recommend adding a TODO comment to improve it in the future.

Comment (uilianries) [inline recipes/libssh/all/conanfile.py:70]:
```suggestion
        # TODO: Add support for mbedtls. Package available in Conan Center already.
        tc.variables["WITH_MBEDTLS"] = False
```

File: recipes/libssh/all/patches/0.10.6-0001-fix-cmake.patch
Comment (uilianries) [inline recipes/libssh/all/patches/0.10.6-0001-fix-cmake.patch:1]: I just checked the upstream and they are actually using CMake targets in the master branch. So it should not be needed in the future.

**Review by uilianries — APPROVED** [review]
LGTM

**Pull request merged**